File: app/src/main/java/info/nightscout/androidaps/plugins/pump/common/defs/DoseStepSize.java
package info.nightscout.androidaps.plugins.pump.common.defs;

/**
 * Created by andy on 02/05/2018.
 */

public enum DoseStepSize {

    ComboBasal( //
            new DoseStepSizeEntry(0f, 1f, 0.01f), //
            new DoseStepSizeEntry(1f, 10f, 0.05f), //
            new DoseStepSizeEntry(10f, Double.MAX_VALUE, 0.1f)), //

    InsightBolus(
            new DoseStepSizeEntry(0f, 2f, 0.05f), //
            new DoseStepSizeEntry(2f, 5f, 0.1f), //
            new DoseStepSizeEntry(5f, 10f, 0.2f), //
            new DoseStepSizeEntry(10f, Double.MAX_VALUE, 0.5f)),

    InsightBasal(
            new DoseStepSizeEntry(0f, 5f, 0.01f),
            new DoseStepSizeEntry(5f, Double.MAX_VALUE, 0.1f)),

    MedtronicVeoBasal( //
            new DoseStepSizeEntry(0f, 1f, 0.025f), //
            new DoseStepSizeEntry(1f, 10f, 0.05f), //
            new DoseStepSizeEntry(10f, Double.MAX_VALUE, 0.1f)), //
    ;


    DoseStepSizeEntry[] entries;


    DoseStepSize(DoseStepSizeEntry... entries) {
        this.entries = entries;
    }


    public double getStepSizeForAmount(double amount) {
        for (DoseStepSizeEntry entry : entries) {
            if (entry.from <= amount && entry.to > amount)
                return entry.value;
        }

        // should never come to this
        return entries[entries.length - 1].value;
    }


    public String getDescription() {
        StringBuilder sb = new StringBuilder();

        for (DoseStepSizeEntry entry : entries) {

            sb.append(entry.value);
            sb.append(" {");
            sb.append(entry.from);
            sb.append("-");

            if (entry.to == Double.MAX_VALUE) {
                sb.append("~}");
            } else {
                sb.append(entry.to);
                sb.append("}, ");
            }
        }

        return sb.toString();
    }


    static class DoseStepSizeEntry {
        double from;
        double to;
        double value;

        // to = this value is not included, but would actually mean <, so for rates between 0.025-0.975 u/h, we would have [from=0, to=10]
        DoseStepSizeEntry(double from, double to, double value) {
            this.from = from;
            this.to = to;
            this.value = value;
        }
    }

}

File: core/data/src/main/kotlin/app/aaps/core/data/pump/defs/DoseStepSize.kt
package app.aaps.core.data.pump.defs

import java.util.Locale

enum class DoseStepSize(private val entries: Array<DoseStepSizeEntry>) {

    ComboBasal(
        arrayOf(
            DoseStepSizeEntry(0.0, 1.0, 0.01),
            DoseStepSizeEntry(1.0, 10.0, 0.05),
            DoseStepSizeEntry(10.0, Double.MAX_VALUE, 0.1)
        )
    ),
    InsightBolus(
        arrayOf(
            DoseStepSizeEntry(0.0, 2.0, 0.05),
            DoseStepSizeEntry(2.0, 5.0, 0.1),
            DoseStepSizeEntry(5.0, 10.0, 0.2),
            DoseStepSizeEntry(10.0, Double.MAX_VALUE, 0.5)
        )
    ),
    InsightBasal(
        arrayOf(
            DoseStepSizeEntry(0.0, 5.0, 0.01),
            DoseStepSizeEntry(5.0, Double.MAX_VALUE, 0.1)
        )
    ),
    MedtronicVeoBasal(
        arrayOf(
            DoseStepSizeEntry(0.0, 1.0, 0.025),
            DoseStepSizeEntry(1.0, 10.0, 0.05),
            DoseStepSizeEntry(10.0, Double.MAX_VALUE, 0.1)
        )
    ),
    YpsopumpBasal(
        arrayOf(
            DoseStepSizeEntry(0.0, 1.0, 0.01),
            DoseStepSizeEntry(1.0, 2.0, 0.02),
            DoseStepSizeEntry(2.0, 15.0, 0.1),
            DoseStepSizeEntry(15.0, 40.0, 0.5)
        )
    ),
    Apex(
        arrayOf(
            DoseStepSizeEntry(0.0, 1.0, 0.025),
            DoseStepSizeEntry(1.0, 2.0, 0.05),
            DoseStepSizeEntry(2.0, Double.MAX_VALUE, 0.1)
        )
    );

    fun getStepSizeForAmount(amount: Double): Double {
        for (entry in entries)
            if (entry.from <= amount && entry.to > amount) return entry.value

        // should never come to this
        return entries[entries.size - 1].value
    }

    val description: String
        get() = StringBuilder().also { sb ->
            var first = true
            for (entry in entries) {
                if (first) first = false else sb.append(", ")

                sb.append(String.format(Locale.ENGLISH, "%.3f", entry.value))
                    .append(" {")
                    .append(String.format(Locale.ENGLISH, "%.3f", entry.from))
                    .append("-")
                if (entry.to == Double.MAX_VALUE) sb.append("~}")
                else sb.append(String.format(Locale.ENGLISH, "%.3f", entry.to)).append("}")
            }
        }.toString()

    // to = this value is not included, but would actually mean <, so for rates between 0.025-0.975 u/h, we would have [from=0, to=10]
    internal class DoseStepSizeEntry(var from: Double, var to: Double, var value: Double)

}